oak/server/rust/oak_runtime/src/runtime/introspect.rs

```diff
@@ -0,0 +1,150 @@
+<script src="https://unpkg.com/d3-graphviz@3.0.0/build/d3-graphviz.js"></script>
+<div id="graph" style="text-align: center;"></div>
+<script>d3.select("#graph").graphviz().renderDot(`
+{}
+`);</script>
+"###,
+        title, graph,
+    )
+}
+
+fn html_wrap(title: &str, body: &str) -> String {
+    format!(
+        r###"
+<!DOCTYPE html>
+<meta charset="utf-8">
+<title>{}</title>
+<body>
+{}
+"###,
+        title, body
+    )
+}
+
+// Look for a path like "/{kind}/{id}" and return the id value.
+fn find_id(path: &str, kind: &str) -> Option<u64> {
+    Regex::new(&format!(r"^/{}/(?P<id>\d+)/?$", kind))
+        .unwrap()
+        .captures(path)?
+        .name("id")
+        .unwrap()
+        .as_str()
+        .parse::<u64>()
+        .ok()
+}
+
+fn handle_request(
+    req: Request<Body>,
+    runtime: Arc<Runtime>,
+) -> Result<Response<Body>, hyper::Error> {
+    if req.method() != Method::GET {
+        let mut not_found = Response::default();
+        *not_found.status_mut() = StatusCode::NOT_FOUND;
+        return Ok(not_found);
+    }
+    // TODO(#672): Shift to a framework.
+    let path = req.uri().path();
+    if path == "/" {
+        return Ok(Response::new(Body::from(html_wrap(
+            "Data Structures",
+            &runtime.html(),
+        ))));
+    } else if path == "/graph" {
+        return Ok(Response::new(Body::from(dot_wrap(
+            "Runtime Graph",
+            &runtime.graph(),
+        ))));
+    } else if let Some(node_id) = find_id(path, "node") {
+        if let Some(body) = runtime.html_for_node(node_id) {
+            return Ok(Response::new(Body::from(html_wrap(
+                &format!("Node {}", node_id),
+                &body,
+            ))));
+        }
+    } else if let Some(channel_id) = find_id(path, "channel") {
+        if let Some(body) = runtime.html_for_channel(channel_id) {
+            return Ok(Response::new(Body::from(html_wrap(
+                &format!("Channel {}", channel_id),
+                &body,
+            ))));
+        }
+    } else if let Some(half_id) = find_id(path, "half") {
+        if let Some(body) = runtime.html_for_half(half_id) {
+            return Ok(Response::new(Body::from(html_wrap(
+                &format!("Channel Half {}", half_id),
+                &body,
+            ))));
+        }
+    }
+    let mut not_found = Response::default();
+    *not_found.status_mut() = StatusCode::NOT_FOUND;
+    Ok(not_found)
+}
+
+async fn make_server(port: u16, runtime: Arc<Runtime>) {
+    // Construct our SocketAddr to listen on...
+    let addr = SocketAddr::from(([127, 0, 0, 1], port));
+    info!("starting introspection server on {:?}", addr);
+
+    // And a MakeService to handle each connection...
+    let make_service = make_service_fn(move |_| {
+        // The `Arc<Runtime>` is moved into this closure, but needs to be cloned
+        // because this closure is called for every connection.
+        let runtime = runtime.clone();
```

Comment:
Perhaps not immediately actionable, but I feel that passing around an actual `Arc` is somewhat too strong, now `Runtime` can never be destroyed while this server is running (which I guess is not a big deal, since they will probably be terminated only when the process is killed). But it kind of feels like we should only pass a weak reference to the runtime here, and perhaps in a few other places related to channel handling. Just something to perhaps discuss separately.

Reply:
Added TODO(#908).